File: tests/test_endpoint.py
"""
LSST Data Management System
See COPYRIGHT file at the top of the source tree.

This product includes software developed by the
LSST Project (http://www.lsst.org/).

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the LSST License Statement and
the GNU General Public License along with this program. If not,
see <http://www.lsstcorp.org/LegalNotices/>.
"""
from __future__ import absolute_import, division, print_function
import unittest

from numpy.testing import assert_allclose, assert_equal
import astshim

import lsst.utils.tests
from lsst.afw.coord import IcrsCoord
import lsst.afw.geom as afwGeom
from lsst.pex.exceptions import InvalidParameterError
from lsst.afw.geom.testUtils import TransformTestBaseClass


class EndpointTestCase(TransformTestBaseClass):

    def setUp(self):
        self.longMessage = True

    def testIcrsCoordEndpoint(self):
        endpoint = afwGeom.IcrsCoordEndpoint()
        self.checkEndpointBasics(
            endpoint=endpoint, pointType=IcrsCoord, nAxes=2)
        self.assertEqual(repr(endpoint), "lsst.afw.geom.IcrsCoordEndpoint()")
        self.assertEqual("{}".format(endpoint), "IcrsCoordEndpoint()")

        for doPermute in (False, True):
            frame = astshim.SkyFrame()
            if doPermute:
                frame.permAxes([2, 1])
                self.assertEqual(frame.lonAxis, 2)
                self.assertEqual(frame.latAxis, 1)
            else:
                self.assertEqual(frame.lonAxis, 1)
                self.assertEqual(frame.latAxis, 2)
            endpoint.normalizeFrame(frame)
            # the normalized frame always has axis in order Lon, Lat
            self.assertEqual(frame.lonAxis, 1)
            self.assertEqual(frame.latAxis, 2)

        badFrame = astshim.Frame(2)
        with self.assertRaises(InvalidParameterError):
            endpoint.normalizeFrame(badFrame)

        newFrame = endpoint.makeFrame()
        self.assertEqual(type(newFrame), astshim.SkyFrame)
        self.assertEqual(newFrame.lonAxis, 1)
        self.assertEqual(newFrame.latAxis, 2)

    def testPoint2Endpoint(self):
        endpoint = afwGeom.Point2Endpoint()
        self.checkEndpointBasics(
            endpoint=endpoint, pointType=afwGeom.Point2D, nAxes=2)
        self.assertEqual(repr(endpoint), "lsst.afw.geom.Point2Endpoint()")
        self.assertEqual("{}".format(endpoint), "Point2Endpoint()")

        # normalize does not check the # of axes
        for n in range(4):
            frame1 = astshim.Frame(n)
            try:
                endpoint.normalizeFrame(frame1)
            except Exception as e:
                self.fail(
                    "endpoint.normalizeFrame(Frame({})) failed with error = {}".format(n, e))
        badFrame = astshim.SkyFrame()
        with self.assertRaises(InvalidParameterError):
            endpoint.normalizeFrame(badFrame)

    def testGenericEndpoint(self):
        for nAxes in (1, 2, 3, 4, 5):
            endpoint = afwGeom.GenericEndpoint(nAxes)
            self.checkEndpointBasics(
                endpoint=endpoint, pointType=list, nAxes=nAxes)
            self.assertEqual(
                repr(endpoint), "lsst.afw.geom.GenericEndpoint({})".format(nAxes))
            self.assertEqual("{}".format(endpoint),
                             "GenericEndpoint({})".format(nAxes))

            newFrame = endpoint.makeFrame()
            self.assertEqual(type(newFrame), astshim.Frame)
            self.assertEqual(newFrame.nAxes, nAxes)

        for nAxes in (-1, 0):
            with self.assertRaises(InvalidParameterError):
                afwGeom.GenericEndpoint(nAxes)

    def checkEndpointBasics(self, endpoint, pointType, nAxes):
        isAngle = pointType == IcrsCoord  # point components are Angles

        baseMsg = "endpoint={}, pointType={}, nAxes={}".format(
            endpoint, pointType, nAxes)

        self.assertEqual(endpoint.nAxes, nAxes, msg=baseMsg)

        # generate enough points to be interesting, but no need to overdo it
        nPoints = 4

        rawData = self.makeRawArrayData(nPoints=nPoints, nAxes=nAxes)
        pointList = endpoint.arrayFromData(rawData)
        self.assertEqual(endpoint.getNPoints(pointList), nPoints, msg=baseMsg)
        if isinstance(endpoint, afwGeom.GenericEndpoint):
            self.assertEqual(len(pointList[0]), nPoints, msg=baseMsg)
            assert_equal(rawData, pointList)
        else:
            self.assertEqual(len(pointList), nPoints, msg=baseMsg)
            for i, point in enumerate(pointList):
                for axis in range(nAxes):
                    msg = "{}, endpoint={}, i={}, point={}".format(
                        baseMsg, endpoint, i, point)
                    if isAngle:
                        desAngle = rawData[axis, i] * afwGeom.radians
                        self.assertAnglesAlmostEqual(
                            point[axis], desAngle, msg=msg)
                    else:
                        self.assertAlmostEqual(
                            point[axis], rawData[axis, i], msg=msg)

        rawDataRoundTrip = endpoint.dataFromArray(pointList)
        self.assertEqual(rawData.shape, rawDataRoundTrip.shape, msg=baseMsg)
        self.assertFloatsAlmostEqual(rawData, rawDataRoundTrip, msg=baseMsg)

        pointData = self.makeRawPointData(nAxes=nAxes)
        point = endpoint.pointFromData(pointData)
        self.assertEqual(type(point), pointType, msg=baseMsg)
        for axis in range(nAxes):
            msg = "{}, axis={}".format(baseMsg, axis)
            if isAngle:
                desAngle = pointData[axis] * afwGeom.radians
                self.assertAnglesAlmostEqual(point[axis], desAngle, msg=msg)
            else:
                self.assertAlmostEqual(point[axis], pointData[axis], msg=msg)

        pointDataRoundTrip = endpoint.dataFromPoint(point)
        assert_allclose(pointData, pointDataRoundTrip, err_msg=baseMsg)

    def testEndpointEquals(self):
        """Test Endpoint == Endpoint
        """
        for i1, point1 in enumerate(self.makeEndpoints()):
            for i2, point2 in enumerate(self.makeEndpoints()):
                if i1 == i2:
                    self.assertTrue(point1 == point2)
                    self.assertFalse(point1 != point2)
                else:
                    self.assertFalse(point1 == point2)
                    self.assertTrue(point1 != point2)


class MemoryTester(lsst.utils.tests.MemoryTestCase):
    pass


def setup_module(module):
    lsst.utils.tests.init()


if __name__ == "__main__":
    lsst.utils.tests.init()
    unittest.main()

File: tests/test_endpoint.cc
// -*- LSST-C++ -*-

/*
 * LSST Data Management System
 * See COPYRIGHT file at the top of the source tree.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program. If not,
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */

#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE EndpointCpp

#include "boost/test/unit_test.hpp"

#include <string>
#include <vector>

#include "lsst/afw/geom/Endpoint.h"

using namespace std;

/* C++ unit tests for Endpoint.
 *
 * In addition to tests of C++-only functionality, this test suite includes
 * tests that need custom subclasses of standard endpoints. While these could
 * be written in Python, allowing Python classes to inherit from pybind11
 * wrappers requires a lot of extra support code.
 *
 * See test_endpoint.py for remaining unit tests.
 */
namespace lsst {
namespace afw {
namespace geom {

namespace {
/**
 * A SpherePointEndpoint whose sole purpose is to be a nontrivial subclass
 * of SpherePointEndpoint.
 */
class GratuitousSpherePointEndpoint : public SpherePointEndpoint {
    // Default constructor is adequate
};

/**
 * An SpherePointEndpoint that adds equality-relevant fields to SpherePointEndpoint.
 */
class CoolSpherePointEndpoint : public SpherePointEndpoint {
public:
    CoolSpherePointEndpoint(string const &coolness) : SpherePointEndpoint(), _coolness(coolness) {}
    string getCoolness() const noexcept { return _coolness; }

    bool operator==(BaseEndpoint const &other) const noexcept override {
        if (!SpherePointEndpoint::operator==(other)) {
            return false;
        } else {
            // Guaranteed not to throw by BaseEndpoint::operator==
            auto coolOther = dynamic_cast<CoolSpherePointEndpoint const &>(other);
            return _coolness == coolOther._coolness;
        }
    }

private:
    string const _coolness;
};
}  // namespace

/// Test whether Endpoint equality treats subclasses as non-substitutable.
BOOST_AUTO_TEST_CASE(EndpointEqualsNotPolymorphic) {
    SpherePointEndpoint superclass;
    GratuitousSpherePointEndpoint subclass;

    // Test both SpherePointEndpoint::operator== and GratuitousSpherePointEndpoint::operator==
    BOOST_TEST(superclass != subclass);
    BOOST_TEST(subclass != superclass);
    BOOST_TEST(!(superclass == subclass));
    BOOST_TEST(!(subclass == superclass));
}

string printEquals(bool equal, string const &name1, string const &name2) {
    return name1 + (equal ? " == " : " != ") + name2;
}

/**
 * Test whether Endpoint equality is symmetric and transitive even in the
 * presence of subclasses that add value fields.
 */
BOOST_AUTO_TEST_CASE(EndpointEqualsAlgebraic) {
    using namespace std::string_literals;

    vector<std::shared_ptr<SpherePointEndpoint const>> endpoints = {
            make_shared<CoolSpherePointEndpoint>("Very cool"s),
            make_shared<CoolSpherePointEndpoint>("Slightly nifty"s),
            make_shared<CoolSpherePointEndpoint>("Very cool"s), make_shared<SpherePointEndpoint>()};

    BOOST_TEST(*(endpoints[0]) != *(endpoints[1]));
    BOOST_TEST(*(endpoints[1]) != *(endpoints[0]));
    BOOST_TEST(*(endpoints[0]) == *(endpoints[2]));
    BOOST_TEST(*(endpoints[2]) == *(endpoints[0]));

    for (auto const endpoint1 : endpoints) {
        for (auto const endpoint2 : endpoints) {
            // == and != must always disagree
            if (*endpoint1 == *endpoint2) {
                BOOST_TEST(!(*endpoint1 != *endpoint2));
            } else {
                BOOST_TEST(*endpoint1 != *endpoint2);
            }

            for (auto const endpoint3 : endpoints) {
                bool const equals12 = *endpoint1 == *endpoint2;
                bool const equals13 = *endpoint1 == *endpoint3;
                bool const equals23 = *endpoint2 == *endpoint3;

                // Test point1 == point2 && point1 == point3 => point2 == point3
                // Iteration covers all permutations, so no need to test, for example,
                //     point1 == point3 && point2 == point3 => point1 == point2
                bool const transitive = !equals12 || !equals13 || equals23;
                BOOST_TEST(transitive, "Not transitive: " << printEquals(equals12, "1", "2") << " "
                                                          << printEquals(equals13, "1", "3") << " "
                                                          << printEquals(equals23, "2", "3"));
            }
        }
    }
}

}  // namespace geom
}  // namespace afw
}  // namespace lsst
